File: schemas/event-management-ticketing/drizzle/venues.ts
// venues: Physical, virtual, or hybrid locations where events take place.
// See README.md for full design rationale.

import {
  pgTable,
  pgEnum,
  uuid,
  text,
  numeric,
  integer,
  boolean,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { users } from "../../auth-rbac/drizzle/users";

export const venueTypeEnum = pgEnum("venue_type", [
  "physical",
  "virtual",
  "hybrid",
]);

export const venues = pgTable(
  "venues",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    name: text("name").notNull(),
    slug: text("slug").notNull().unique(),
    description: text("description"),
    type: venueTypeEnum("type").notNull().default("physical"),
    addressLine1: text("address_line1"),
    addressLine2: text("address_line2"),
    city: text("city"),
    state: text("state"),
    postalCode: text("postal_code"),
    country: text("country"),
    latitude: numeric("latitude"),
    longitude: numeric("longitude"),
    virtualUrl: text("virtual_url"),
    virtualPlatform: text("virtual_platform"),
    capacity: integer("capacity"),
    timezone: text("timezone").notNull(),
    phone: text("phone"),
    email: text("email"),
    websiteUrl: text("website_url"),
    isActive: boolean("is_active").notNull().default(true),
    createdBy: uuid("created_by")
      .notNull()
      .references(() => users.id, { onDelete: "restrict" }),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("idx_venues_type").on(table.type),
    index("idx_venues_city_state").on(table.city, table.state),
    index("idx_venues_is_active").on(table.isActive),
    index("idx_venues_created_by").on(table.createdBy),
  ],
);

File: schemas/event-management-ticketing/spacetimedb/venues.rs
// venues: Physical, virtual, or hybrid locations where events take place.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum VenueType {
    Physical, // type: String
    Virtual,
    Hybrid,
}

#[spacetimedb::table(name = venues, public)]
pub struct Venue {
    #[primary_key]
    pub id: String, // UUID

    pub name: String,

    #[unique]
    pub slug: String,

    pub description: Option<String>,

    #[index(btree)]
    pub venue_type: VenueType,

    pub address_line1: Option<String>,
    pub address_line2: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub postal_code: Option<String>,
    pub country: Option<String>,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    pub virtual_url: Option<String>,
    pub virtual_platform: Option<String>,
    pub capacity: Option<i32>,
    pub timezone: String,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub website_url: Option<String>,
    pub is_active: bool,

    #[index(btree)]
    pub created_by: String, // UUID — FK → users.id (restrict delete)

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
    // Composite index: (city, state)
}
